internal/driver/readhandler.go

```diff
@@ -24,23 +46,86 @@ func (d *Driver) HandleReadCommands(deviceName string, protocols map[string]mode
 
 	d.Logger.Debugf("Driver.HandleReadCommands: protocols: %v resource: %v attributes: %v", protocols, reqs[0].DeviceResourceName, reqs[0].Attributes)
 
-	// create device client and open connection
-	endpoint, err := config.FetchEndpoint(protocols)
+	endpoints, err := opcua.GetEndpoints(d.serviceConfig.OPCUAServer.Endpoint)
 	if err != nil {
-		return nil, err
+		d.Logger.Error("OPC GetEndpoints: %w", err)
+	}
+	credentials, err := getCredentials(d.serviceConfig.OPCUAServer.CredentialsPath)
+	if err != nil {
+		d.Logger.Error("getCredentials: %w", err)
+	}
+
+	username := credentials.Username
+	password := credentials.Password
+	policy := ua.SecurityPolicyURIBasic256Sha256
+	mode := ua.MessageSecurityModeSignAndEncrypt
+
+	ep := opcua.SelectEndpoint(endpoints, policy, mode)
+	c, err := generateCert() // This is where you generate the certificate
+	if err != nil {
+		d.Logger.Error("generateCert: %w", err)
```

Comment:
Same

Reply:
made some refactoring work, still potential but at least something for now :)

Reply:
For me here is still too much one line variable names, still some return on error missing and for me the try to get some unittest.

Reply:
unit tests still dont work, generally in this go repo. That is a general issue we would need to tackle separately